src/main/scala/nl/knaw/dans/api/sword2/DepositHandler.scala

```diff
@@ -152,51 +154,117 @@ object DepositHandler {
+
     val bag = getBagFromDir(bagitDir)
     val validationResult: SimpleResult = bag.verifyValid
-    if (validationResult.isSuccess) Success(Unit)
-    else Failure(new InvalidDepositException(id, validationResult.messagesToString))
+
+    if (fetchTextsBagStore.isEmpty) {
+      if (validationResult.isSuccess) Success(Unit)
+      else Failure(InvalidDepositException(id, validationResult.messagesToString))
+    }
+    else {
+      if (validationResult.isSuccess) log.warn(s"[$id] There is a fetch.txt file, while all the files are present in the bag.")
+
+      val otherThanMissingPayloadFilesMessages = validationResult.getSimpleMessages.listIterator().asScala.filterNot(msg =>
+        msg.getCode == CompleteVerifier.CODE_PAYLOAD_MANIFEST_CONTAINS_MISSING_FILE)
+      if (otherThanMissingPayloadFilesMessages.nonEmpty)
+        Failure(InvalidDepositException(id, otherThanMissingPayloadFilesMessages.mkString))
+      else {
+        val missingPayloadFiles = validationResult.getSimpleMessages.listIterator().asScala.toList.flatMap(msg => msg.getObjects.asScala.toList)
+        val fetchTextFilesBagStore = fetchTextsBagStore.map(item => item.getFilename)
+        val missingFilesNotInFetchText = missingPayloadFiles.filterNot(file => fetchTextFilesBagStore.contains(file))
+        if (missingFilesNotInFetchText.nonEmpty)
+          Failure(InvalidDepositException(id, s"Missing payload files not in the fetch.txt: $missingFilesNotInFetchText.mkString."))
+        else {
+          validateChecksumsFetchtextFiles(bag, fetchTextsBagStore)
+        }
+      }
+    }
+  }
+
+  private def resolveFetchTxtFiles(bagitDir: File, fetchTexts: List[FetchTxt.FilenameSizeUrl])(implicit id: String): Unit =  {
+    if (fetchTexts.nonEmpty){
+      log.debug(s"[$id] Resolving files in fetch.txt, those referring outside the bag store.")
+      fetchTexts.foreach(item =>
+        Using.urlInputStream(new URL(item.getUrl)).foreach(src =>
+          Files.copy(src, Paths.get(bagitDir.getAbsolutePath, item.getFilename))))
+    }
+  }
+
+  private def validateChecksumsFetchtextFiles(bag: Bag, fetchTexts: List[FetchTxt.FilenameSizeUrl]) (implicit id: String) = {
+    log.debug(s"[$id] Validating the checksums of those files in fetch.txt, that refer to the bag store.")
+
+    val fetchTextFiles = fetchTexts.map(item => item.getFilename)
+    val checksums = bag.getPayloadManifests.asScala.toList.flatMap(manifest => manifest.asScala).filter(fileChecksumTuple => fetchTextFiles.contains(fileChecksumTuple._1))
+    val urls = fetchTexts.map(file => file.getFilename -> file.getUrl)
+
+    val checksumMapping: List[(String, String, String)] =
+      checksums.map { case (file_a, checksum) =>
+        urls.find { case (file_b, url) => file_a == file_b }
+          .map { case (_, url) => (file_a, checksum, url) }
+            .getOrElse(throw InvalidDepositException(id, s"Checksum validation failed: missing Payload Manifest file $file_a not found in the fetch.txt."))
```

Comment:
Why not? The exception will be caught in `finalizeDeposit` method, where all the error messages are logged.  If I would make it `Try/Success/Failure`, then I would need to write more code to dig up the original error message.